src/System.IO.FileSystem/src/System/IO/DosMatcher.cs

```diff
@@ -0,0 +1,307 @@
+                {
+                    // Handle the special case of a single starting *, which essentially means "ends with"
+
+                    // If the name doesn't have enough characters to match the remaining expression, it can't be a match.
+                    if (name.Length < expression.Length - 1)
+                        return false;
+
+                    // See if we end with the expression (minus the *, of course)
+                    return name.EndsWithOrdinal(expression.AsReadOnlySpan().Slice(1), ignoreCase);
+                }
+            }
+
+            int nameOffset = 0;
+            int expressionOffset;
+
+            int priorMatch;
+            int currentMatch;
+            int priorMatchCount;
+            int matchCount = 1;
+
+            char nameChar = '\0';
+            char expressionChar;
+
+            Span<int> temp = stackalloc int[0];
+            Span<int> priorMatches = stackalloc int[16];
+            Span<int> currentMatches = stackalloc int[16];
+
+            int maxState = expression.Length * 2;
+            int currentState;
+            bool nameFinished = false;
+
+            while (!nameFinished)
+            {
+                if (nameOffset < name.Length)
+                {
+                    // Not at the end of the name. Grab the current character and move the offset forward.
+                    nameChar = name[nameOffset++];
+                }
+                else
+                {
+                    // At the end of the name. If the expression is exhausted, exit.
+                    if (priorMatches[matchCount - 1] == maxState)
+                        break;
+
+                    nameFinished = true;
+                }
+
+                // Now, for each of the previous stored expression matches, see what
+                // we can do with this name character.
+                priorMatch = 0;
+                currentMatch = 0;
+                priorMatchCount = 0;
+
+                while (priorMatch < matchCount)
+                {
+                    // We have to carry on our expression analysis as far as possible for each
+                    // character of name, so we loop here until the expression stops matching.
+
+                    expressionOffset = (priorMatches[priorMatch++] + 1) / 2;
+
+                    while (expressionOffset < expression.Length)
+                    {
+                        currentState = expressionOffset * 2;
+                        expressionChar = expression[expressionOffset];
+
+                        // We may be about to exhaust the local space for matches,
+                        // so we have to reallocate if this is the case.
+                        if (currentMatch >= currentMatches.Length - 2)
+                        {
+                            int newSize = currentMatches.Length * 2;
+                            temp = new Span<int>(new int[newSize]);
```

Comment:
Nit: there's an implicit cast from array to span... this can just be:
```C#
temp = new int[newSize];
```

Reply:
Thanks